pkg/client/clients.go

```diff
@@ -27,13 +27,19 @@ import (
 	coreInformerV1 "k8s.io/client-go/informers/core/v1"
 	schedulingInformerV1 "k8s.io/client-go/informers/scheduling/v1"
 	storageInformerV1 "k8s.io/client-go/informers/storage/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding"
 
 	"github.com/apache/yunikorn-k8shim/pkg/conf"
 	"github.com/apache/yunikorn-k8shim/pkg/log"
 	"github.com/apache/yunikorn-scheduler-interface/lib/go/api"
 )
 
+type informer interface {
```

Comment:
Maybe we just add a new method to `Clients` to return all informers. For example:
```go
func (c *Clients) informers() []cache.SharedIndexInformer
```
The true impl is to collect informers from all resource informers. WDYT

Reply:
Is it better to call  `Informer() cache.SharedIndexInformer` every time we need it?

Reply:
> Is it better to call Informer() cache.SharedIndexInformer every time we need it?

It seems both `WaitForSync` and `Run` are not hot methods, right? If so, creating the new collection should not a issue about performance regression I think.

Reply:
I came across this comment in the source code:
```
// Always prefer using an informer factory to get a shared informer instead of getting an independent
// one. This reduces memory footprint and number of connections to the server.
```
There doesn't appear to be any issue with it currently.
But I think we should avoid any potential hidden bugs in the codebase.

Reply:
> But I think we should avoid any potential hidden bugs in the codebase.

The cost we are taking about is "creating a slice" rather than calling `Informer()`. We do call `Informer()` every time in `WaitForSync` and `Run`, right? Hence, if `WaitForSync` and `Run` are not hot methods, the cost of creating slice won't be a issue to me.

Reply:
I got it

Reply:
I think this is getting a little out of hand. This is initialization code to handle common Kubernetes objects. We don't want to expose slices, or other internal implementation details. This code is also not hot, as it gets called once during execution. We should favor simplicity here. I think this new implementation is overly complex.

Reply:
I've updated the pr based on @chia7712's comment.
I would add test to pass the CodeCov later.

Reply:
I looked at using something like slices for storing all the different informers. Due to the differences of the listers between the informers that becomes a slice of `interface{}` That then triggers type casting and all kinds of other handling and checks which is worse than having the separate variables and maintaining the two sets. Just storing the `Informer()` and not the `Lister()` for the informers does not make sense for me. The lister is normally accessed outside of startup the informer 

Any informer defined should be started and stopped. To not regress we could have a simple unit test that checks that all informers defined in the type are also returned in the `getInformers()` call. If someone adds a new informer we can then ensure that they are updated in both places

Reply:
> Any informer defined should be started and stopped. To not regress we could have a simple unit test that checks that all informers defined in the type are also returned in the getInformers() call. If someone adds a new informer we can then ensure that they are updated in both places

agree to that "started" and "stopped" are the most important here. It means `Clients` must have all "created" informers. Maybe we can add a small helper to collect all created informers into a slice, and then we pass the slice to `Clients`. For example:

```go
type hasInformer interface {
	Informer() cache.SharedIndexInformer
}

func save[T hasInformer](n T, collector *[]cache.SharedIndexInformer) T {
	*collector = append(*collector, n.Informer())
	return n
}
```
We use a little generic here to avoid casting, and all created informers should be wrapped by `save`. For example:
```go
	var allInformers []cache.SharedIndexInformer

	nodeInformer := save(informerFactory.Core().V1().Nodes(), &allInformers)
	podInformer := save(informerFactory.Core().V1().Pods(), &allInformers)
```

Finally, `allInformers` have all created informers, and `Clients` will use it to start/stop all informers.

@wilfred-s WDYT?